numba/tests/test_literal_dispatch.py

```diff
@@ -0,0 +1,254 @@
+        # Check that literally accepts inlined literal
+        @njit
+        def foo(a, b):
+            v = 1000
+            return a + literally(v) + literally(b)
+
+        got = foo(1, 2)
+        self.assertEqual(got, foo.py_func(1, 2))
+
+        @njit
+        def bar():
+            a = 100
+            b = 9
+            return foo(a=b, b=a)
+
+        got = bar()
+        self.assertEqual(got, bar.py_func())
+
+    def test_aliased_variable(self):
+        @njit
+        def foo(a, b, c):
+            def closure(d):
+                return literally(d) + 10 * inner(a, b)
+            # The inlining of the closure will create an alias to c
+            return closure(c)
+
+        @njit
+        def inner(x, y):
+            return x + literally(y)
+
+        kwargs = dict(a=1, b=2, c=3)
+        got = foo(**kwargs)
+        expect = (lambda a, b, c: c + 10 * (a + b))(**kwargs)
+        self.assertEqual(got, expect)
+        [(type_a, type_b, type_c)] = foo.signatures
+        self.assertNotIsInstance(type_a, types.Literal)
+        self.assertIsInstance(type_b, types.Literal)
+        self.assertEqual(type_b.literal_value, 2)
+        self.assertIsInstance(type_c, types.Literal)
+        self.assertEqual(type_c.literal_value, 3)
+
+    def test_overload_explicit(self):
+        # This test represents a more controlled usage with ensuring literal
+        # typing for an argument.
+        def do_this(x, y):
+            return x + y
+
+        @overload(do_this)
+        def ov_do_this(x, y):
+            SentryLiteralArgs(['x']).for_function(ov_do_this).bind(x, y)
+            return lambda x, y: x + y
+
+        @njit
+        def foo(a, b):
+            return do_this(a, b)
+
+        a = 123
+        b = 321
+        r = foo(a, b)
+        self.assertEqual(r, a + b)
+        [type_a, type_b] = foo.signatures[0]
+        self.assertIsInstance(type_a, types.Literal)
+        self.assertEqual(type_a.literal_value, a)
+        self.assertNotIsInstance(type_b, types.Literal)
+
+    def test_overload_implicit(self):
+        # This test represents the preferred usage style for using literally
+        # in overload. Here, literally() is used inside the "implementation"
+        # function of the overload.
+        def do_this(x, y):
+            return x + y
+
+        @njit
+        def hidden(x, y):
+            return literally(x) + y
+
+        @overload(do_this)
+        def ov_do_this(x, y):
+            if isinstance(x, types.Integer):
+                # At this point, `x` can be a literal or not
+                return lambda x, y: hidden(x, y)
+
+        @njit
+        def foo(a, b):
+            return do_this(a, b)
+
+        a = 123
+        b = 321
+        r = foo(a, b)
+        self.assertEqual(r, a + b)
+        [type_a, type_b] = foo.signatures[0]
+        self.assertIsInstance(type_a, types.Literal)
+        self.assertEqual(type_a.literal_value, a)
+        self.assertNotIsInstance(type_b, types.Literal)
+
+
```

Comment:
It got there because:

- the Literal'ed version would raise the ValueError
- the overload resolution retries with a Unliteral'ed version
- that triggers a new ForceLiteralArg signal
- that asks for the already Literal type to be wrapped in a Literal type.

If we get rid of the assertion, it just loops forever trying to compile the function.
I think I will make it abort in place of that assertion.

Reply:
I've added a test to ensure that the compilation is aborted with a msg about repeated literal typing request.

Reply:
great, thanks! and thanks for the explanation too.